File: Python/0337-house-robber-iii.py
# First attempt: Time limit exceeded
# Time: O(exponential)
# Space: O(h)
# Post order traversal
class Solution:
    def rob(self, root: TreeNode) -> int:
            if not root:
                return 0

            val = 0

            if root.left:
                val += self.rob(root.left.left) + self.rob(root.left.right)

            if root.right:
                val += self.rob(root.right.left) + self.rob(root.right.right)

            return max(val + root.val, self.rob(root.left) + self.rob(root.right))



# DP solution (memoization): Time limit exceeded
# Time: O(n)
# Space: O(n), stack cost of recursion not counted
class Solution:
    def rob(self, root: TreeNode) -> int:

            def robSub(root):
                if not root:
                    return 0
                if root in node_map:
                    return node_map[root]

                val = 0

                if root.left:
                    val += self.rob(root.left.left) + self.rob(root.left.right)

                if root.right:
                    val += self.rob(root.right.left) + self.rob(root.right.right)

                val = max(val + root.val, self.rob(root.left) + self.rob(root.right))
                node_map[root] = val
                return val

            node_map = {}
            return robSub(root)

# Notes:
# In first solution, we only considered the aspect of "optimal substructure",
# but think little about the possibilities of overlapping of the subproblems.
# For example, to obtain rob(root), we need rob(root.left), rob(root.right),
# rob(root.left.left), rob(root.left.right), rob(root.right.left),
# rob(root.right.right); but to get rob(root.left), we also need
# rob(root.left.left), rob(root.left.right), similarly for rob(root.right).
# The naive solution above computed these subproblems repeatedly, which
# resulted in bad time performance. Now if you recall the two conditions for
# dynamic programming (DP): "optimal substructure" + "overlapping of subproblems",
# and optimization problem which it is as we have to find maximum.
# we actually have a DP problem. A naive way to implement DP here is to
# use a hash map to record the results for visited subtrees.



# Greedy solution: Accepted
# Time: O(n)
# Space: O(1), stack cost of recursion not counted
# O(h) including stack
class Solution:
    def rob(self, root: TreeNode) -> int:

            def robSub(root):
                if not root:
                    return (0, 0)
                left = robSub(root.left)
                right = robSub(root.right)
                # root is not included
                a = max(left[0], left[1]) + max(right[0], right[1])
                # root not chosen in both left and right subtree
                b = root.val + left[0] + right[0]
                return (a, b)
            return max(robSub(root))

""""
Here, we no longer have repeated subproblems.

What makes it greedy?
Greedy-choice property (a local optical choice at each step, refer link),
Optimal substructure
and it is an Optimization problem.

robSub return an array of two elements, the first element of which denotes
the maximum amount of money that can be robbed if root is not robbed, while
the second element signifies the maximum amount of money robbed if it is robbed.

Ref:https://leetcode.com/problems/house-robber-iii/discuss/79330/Step-by-step-tackling-of-the-problem
"""

File: C++/0337-house-robber-iii.cpp
// Tags: Tree DFS DP Amazon
// Time: O(n)
// Space: O(n), due to map
// Memoization solution
// Check Python version for comments
class Solution {
    unordered_map<TreeNode*, int> memo;
    // maximum amount that can be robbed from tree rooted at 'root'
    int dp(TreeNode* root) {
        if (!root)
            return 0;
        if (memo.count(root))
            return memo[root];
        int val = root->val;
        if (root->left)
            val += dp(root->left->left) + dp(root->left->right);
        if (root->right)
            val += dp(root->right->left) + dp(root->right->right);
        memo[root] = max(val, dp(root->left) + dp(root->right));
        return memo[root];
    }
public:
    int rob(TreeNode* root) {
        return dp(root);
    }
};
/*
dp[NULL] = 0;
dp[node] = max(node->val + sigma over (dp[g]), sigma over (dp[c]))
    where c denotes children of node and g denotes grandchildren of g
*/



// Time: O(n)
// Space: O(1) w/o stack, O(h) w/ stack
// Greedy solution
// Check Python version for comments
class Solution {
    // maximum amount of money that can be robbed from tree rooted at 'root'
    // if 'root' is not robbed, if 'root' is robbed
    vector<int> robSum(TreeNode* root) {
        if (!root)
            return {0, 0};
        auto left = robSum(root->left);
        auto right = robSum(root->right);
        // Note that a is not left[1] + right[1]
        // if root is not robbed, children may or may not be robbed
        // or may be just one of them is robbed
        int a = max(left[0], left[1]) + max(right[0], right[1]);
        // if root is robbed, children can't be robbed
        int b = root->val + left[0] + right[0];
        return {a, b};
    }
public:
    int rob(TreeNode* root) {
        auto res = robSum(root);
        return max(res[0], res[1]);
    }
};
// No longer have repeated subproblems
